extensions/rv32im/circuit/src/adapters/alu.rs

```diff
@@ -178,56 +179,85 @@ impl<F: PrimeField32, CTX, const LIMB_BITS: usize> AdapterTraceStep<F, CTX>
 
     #[inline(always)]
     fn read(
+        &self,
         memory: &mut TracingMemory,
         instruction: &Instruction<F>,
         adapter_row: &mut [F],
     ) -> Self::ReadData {
         let &Instruction { b, c, d, e, .. } = instruction;
+
         debug_assert_eq!(d.as_canonical_u32(), RV32_REGISTER_AS);
-        let adapter_row: &mut Rv32BaseAluAdapterCols<F> = adapter_row.borrow_mut();
-        let rs1_idx = b.as_canonical_u32();
-        let rs1 = tracing_read_reg(
-            memory,
-            rs1_idx,
-            (&mut adapter_row.rs1_ptr, &mut adapter_row.reads_aux[0]),
+        debug_assert!(
+            e.as_canonical_u32() == RV32_REGISTER_AS || e.as_canonical_u32() == RV32_IMM_AS
         );
-        let rs2 = tracing_read_reg_or_imm(
```

Comment:
why did you get rid of `tracing_read_reg_or_imm`?

Reply:
i had to change the signature of `tracing_read` to not contain the mutable address pointer because sometimes the address pointer is derived from other columns and that made it incompatible with the imm function so i split them